## Импортируем библиотеки

In [1]:
import pandas as pd
from sqlalchemy import create_engine

import dash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output

import seaborn as sns
import math
import numpy as np
from scipy import stats as st

import plotly.express as px
import plotly.graph_objs as go
from plotly.offline import iplot
from plotly import graph_objects as go
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from warnings import simplefilter



## Создаём коннекцию к базе

In [2]:


db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-zen-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string) 

In [3]:
# получаем сырые данные
query = '''
            SELECT * 
            FROM log_raw
        '''
log_raw = pd.io.sql.read_sql(query, con=engine, index_col='event_id')

log_raw.to_csv('log_raw.csv', index=False)

log_raw.head(5) 

age_segment  event  item_id   item_topic item_type  source_id  \
event_id                                                                 
2931262        18-25   show  1793260      Красота    native    8388799   
2931277        18-25   show  7465641     Культура    native    5837959   
2931285        18-25   show   406073      История       url    2335872   
2931247        26-30   show   752665  Путешествия    native    6963489   
2931278        26-30  click  6843081         Юмор       url    5427748   

                source_topic source_type             ts  user_id  
event_id                                                          
2931262              История   publisher  1569349657638  2185026  
2931277              История   publisher  1569349657971  6845238  
2931285           Сад и дача      domain  1569349658408  5244363  
2931247             Здоровье   publisher  1569349657189  7553074  
2931278   Семейные отношения      domain  1569349658054  1317585

## Изучаем данные

In [4]:
list_df = (log_raw,)
for list_df in list_df:
    display('______________info_____________________')
    display(list_df.info())
    display('______________describe_________________')
    display(list_df.describe())
    display('______________head_____________________')
    display(list_df.head(10))
    display('______________sample___________________')
    display(list_df.sample(20))
    display('______________tail_____________________')
    display(list_df.tail(10))
    display('_____________isna______________________')
    display(list_df.isna().sum().reset_index())
    display('_____________duplicated________________')
    display(list_df.duplicated().sum())
    display('__количество пропущенных значений по каждому из столбцов__')
    display(list_df.isnull().mean().reset_index())

'______________info_____________________'

<class 'pandas.core.frame.DataFrame'>
Int64Index: 322391 entries, 2931262 to 4396838
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   age_segment   322391 non-null  object
 1   event         322391 non-null  object
 2   item_id       322391 non-null  int64 
 3   item_topic    322391 non-null  object
 4   item_type     322391 non-null  object
 5   source_id     322391 non-null  int64 
 6   source_topic  322391 non-null  object
 7   source_type   322391 non-null  object
 8   ts            322391 non-null  int64 
 9   user_id       322391 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 27.1+ MB


None

'______________describe_________________'

item_id     source_id            ts       user_id
count  3.223910e+05  3.223910e+05  3.223910e+05  3.223910e+05
mean   4.907860e+06  4.974883e+06  1.569351e+12  4.988528e+06
std    2.918019e+06  2.865432e+06  3.253844e+05  2.898068e+06
min    1.160000e+02  3.484000e+03  1.569350e+12  1.600000e+01
25%    2.306626e+06  2.613120e+06  1.569351e+12  2.481560e+06
50%    4.874690e+06  4.862937e+06  1.569351e+12  4.957050e+06
75%    7.508017e+06  7.573313e+06  1.569352e+12  7.512082e+06
max    9.999779e+06  9.999497e+06  1.569352e+12  9.999930e+06

'______________head_____________________'

age_segment  event  item_id       item_topic       item_type  \
event_id                                                                
2931262        18-25   show  1793260          Красота          native   
2931277        18-25   show  7465641         Культура          native   
2931285        18-25   show   406073          История             url   
2931247        26-30   show   752665      Путешествия          native   
2931278        26-30  click  6843081             Юмор             url   
2931239        31-35   show  3803320            Наука             url   
2931294        36-40  click  4995713             Дети          native   
2931260          45+   show  6459474  Полезные советы  narrative_card   
2931266          45+   show   349207          История          native   
2931298        18-25   show  8438488              Шоу          native   

          source_id        source_topic source_type             ts  user_id  
event_id                                                                     
2931262     8388799             История   publisher  1569349657638  2185026  
2931277     5837959             История   publisher  1569349657971  6845238  
2931285     2335872          Сад и дача      domain  1569349658408  5244363  
2931247     6963489            Здоровье   publisher  1569349657189  7553074  
2931278     5427748  Семейные отношения      domain  1569349658054  1317585  
2931239     9321222        Знаменитости      domain  1569349656520  1596831  
2931294     8137788        Знаменитости   publisher  1569349658461  4498137  
2931260     1576480              Россия   publisher  1569349657524  8959444  
2931266     4422076         Путешествия   publisher  1569349657642  8172088  
2931298     2471071                Кино   publisher  1569349658522  4492894

'______________sample___________________'

age_segment  event  item_id        item_topic       item_type  \
event_id                                                                 
4390842        31-35  click  7527215          Здоровье             url   
4058508        31-35   show  8704194             Семья          native   
4253278        31-35   show  5121090             Семья             url   
3689333        41-45   show  3698403         Искусство             url   
3863584        18-25   show  3278426             Наука             url   
3910356        18-25  click  5979333             Наука             url   
4145340        26-30  click  6131330          Подборки             url   
3787289        31-35   show  4907623            Деньги             url   
3112892        31-35   show  2406257          Скандалы          native   
3676064        18-25   show  7320449   Полезные советы             url   
3115256        18-25   show  6651755          Скандалы  narrative_card   
4184475        36-40   view  7012839          Общество          native   
3292997        26-30   show  5966390             Наука          native   
3131728        18-25   show  8612933   Полезные советы             url   
3363242        18-25   show  2309291            Туризм          native   
3918977        31-35   show  6444187            Деньги          native   
3963458        18-25   show  9509706  Интересные факты          native   
3459721        26-30   show  2306626          Рассказы             url   
3135026        26-30   show  1722564   Полезные советы          native   
3180139        18-25   show  2306626          Рассказы             url   

          source_id        source_topic source_type             ts  user_id  
event_id                                                                     
4390842     8851565         Путешествия      domain  1569351641614  2862400  
4058508     2692676        Знаменитости   publisher  1569351520495  9208148  
4253278     5633014     Полезные советы      domain  1569351567940   458011  
3689333     8988354               Семья      domain  1569351438861  1623034  
3863584     8003506         Путешествия      domain  1569351476611  8670222  
3910356     1924501                Дети      domain  1569351486930  8887682  
4145340     5427748  Семейные отношения      domain  1569351540992  2331355  
3787289      561976             Финансы      domain  1569351460027  4737685  
3112892     6470225              Россия   publisher  1569351291751  1689984  
3676064     8482815         Путешествия      domain  1569351436002  7539091  
3115256     6470225              Россия   publisher  1569351292836  7390709  
4184475     2843391            Политика   publisher  1569351550538  3892194  
3292997     4788263          Технологии   publisher  1569351352163  6347148  
3131728     7968940            Политика      domain  1569351299488  8581692  
3363242     5838056                Дети   publisher  1569351369225  5279388  
3918977     6000142  Семейные отношения   publisher  1569351488762  6423828  
3963458     7377897     Полезные советы   publisher  1569351498670  8430720  
3459721     6381816         Путешествия      domain  1569351389764  7809610  
3135026     3158926  Семейные отношения   publisher  1569351300732  3851594  
3180139     6381816         Путешествия      domain  1569351317445  8819430

'______________tail_____________________'

age_segment  event  item_id          item_topic       item_type  \
event_id                                                                   
4396823        18-25   show   373042                Юмор          native   
4396824        18-25   show  7758213               Семья             url   
4396816        26-30  click  6247520              Россия          native   
4396817        26-30   show  6509942           Отношения          native   
4396821        26-30   show  5966390               Наука          native   
4396804        31-35  click  7273256               Наука          native   
4396814        31-35   show  4094446  Женская психология  narrative_card   
4396836        18-25   show  5003354    Интересные факты             url   
4396846        18-25  click    39740           Отношения  narrative_card   
4396838        31-35   show  3317563            Общество             url   

          source_id        source_topic source_type             ts  user_id  
event_id                                                                     
4396823     9612036                Дети   publisher  1569351658854  2650912  
4396824     8651701        Знаменитости      domain  1569351658854  2650912  
4396816     8388799             История   publisher  1569351658740  4576710  
4396817     3205584              Деньги   publisher  1569351658746  1870531  
4396821     4788263          Технологии   publisher  1569351658787  7969727  
4396804     5451530          Сад и дача   publisher  1569351658610  1351271  
4396814     4001238          Психология   publisher  1569351658738  4466455  
4396836     6371686     Полезные советы      domain  1569351659030  4916583  
4396846     9829090  Семейные отношения   publisher  1569351659233  7962286  
4396838     8280856              Деньги      domain  1569351659036  2720789

'_____________isna______________________'

index  0
0   age_segment  0
1         event  0
2       item_id  0
3    item_topic  0
4     item_type  0
5     source_id  0
6  source_topic  0
7   source_type  0
8            ts  0
9       user_id  0

'_____________duplicated________________'

0

'__количество пропущенных значений по каждому из столбцов__'

index    0
0   age_segment  0.0
1         event  0.0
2       item_id  0.0
3    item_topic  0.0
4     item_type  0.0
5     source_id  0.0
6  source_topic  0.0
7   source_type  0.0
8            ts  0.0
9       user_id  0.0

Вывод:
* Данные из БД успешно выгружены в "csv" файл
* Данные содержат 322391  строк и 10 колонок
* Пропусков и дублей нет
* Необходимо преобразовать данные из колонки `ts`


## Подготовка данных

In [5]:
log_raw['event_time'] = pd.to_datetime(log_raw['ts'], unit = 'ms')
log_raw.head()

age_segment  event  item_id   item_topic item_type  source_id  \
event_id                                                                 
2931262        18-25   show  1793260      Красота    native    8388799   
2931277        18-25   show  7465641     Культура    native    5837959   
2931285        18-25   show   406073      История       url    2335872   
2931247        26-30   show   752665  Путешествия    native    6963489   
2931278        26-30  click  6843081         Юмор       url    5427748   

                source_topic source_type             ts  user_id  \
event_id                                                           
2931262              История   publisher  1569349657638  2185026   
2931277              История   publisher  1569349657971  6845238   
2931285           Сад и дача      domain  1569349658408  5244363   
2931247             Здоровье   publisher  1569349657189  7553074   
2931278   Семейные отношения      domain  1569349658054  1317585   

                      event_time  
event_id                          
2931262  2019-09-24 18:27:37.638  
2931277  2019-09-24 18:27:37.971  
2931285  2019-09-24 18:27:38.408  
2931247  2019-09-24 18:27:37.189  
2931278  2019-09-24 18:27:38.054

## Изучим данные

In [6]:
dash_visits = log_raw.groupby(['item_topic', 'source_topic','age_segment','event_time'])\
                                                .agg({'user_id': 'count'}).reset_index()
dash_visits.head()

item_topic source_topic age_segment              event_time  user_id
0     Деньги         Авто       18-25 2019-09-24 18:31:40.031        1
1     Деньги         Авто       18-25 2019-09-24 18:31:56.354        1
2     Деньги         Авто       18-25 2019-09-24 18:32:20.475        1
3     Деньги         Авто       18-25 2019-09-24 18:35:02.718        1
4     Деньги         Авто       18-25 2019-09-24 18:53:58.172        1

In [7]:
log_raw.item_type.unique()

array(['native', 'url', 'narrative_card', 'post'], dtype=object)

In [8]:
log_raw.source_type.unique()

array(['publisher', 'domain'], dtype=object)

Сгруппируем темы источника по способу взаимодействия пользователей с системой:

In [9]:
log_raw_source = log_raw.pivot_table(index='source_topic', columns='event', values='item_topic', aggfunc='count')\
                                                            .reset_index().sort_values(by='show', ascending=False)
log_raw_source

event        source_topic  click   show  view
20     Семейные отношения   2879  29678  2057
17                 Россия   2251  27161  1608
12        Полезные советы   1957  24424  2156
15            Путешествия   1289  22417  1368
5            Знаменитости   2043  21211  1639
9                    Кино   1579  17503  1802
2                    Дети   1430  13241  1176
8                 История   1101  12885  1190
21                  Семья   1207  12072  1186
4                Здоровье   1071  11363   922
11                 Одежда    938  10452   934
0                    Авто    563   8761   630
7               Искусство    507   7179   665
13               Политика    612   6925    85
18             Сад и дача    460   6791   505
3                     Еда    409   6213   545
19             Сделай сам    368   5747   406
16                 Ремонт    375   5224   332
14             Психология    471   4924   534
22                  Спорт    380   4778   273
1                  Деньги    387   4574   393
6               Интерьеры    212   3359   188
24             Технологии    186   3322   114
23          Строительство    167   2768   176
10                 Музыка    220   2621   133
25                Финансы    173   2376   170

In [10]:
fig = go.Figure()
fig.add_trace(go.Bar(
    y=log_raw_source['click'],
    x=log_raw_source['source_topic'],
    name='Карточка отображена для пользователя',
    marker_color='#f0932b'
))
fig.add_trace(go.Bar(
    y=log_raw_source['show'],
    x=log_raw_source['source_topic'],
    name='Пользователь кликнул на карточку',
    marker_color='#686de0'
))

fig.add_trace(go.Bar(
    y=log_raw_source['view'],
    x=log_raw_source['source_topic'],
    name='Пользователь просмотрел статью карточки',
    marker_color='#22a6b3'
))

fig.update_layout(barmode='group', xaxis_tickangle=-45,
        title={
        'text': 'Способы взаимодействия пользователей с системой',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.show()

Сгруппируем статьи по способу взаимодействия пользователей с системой:

In [11]:
log_raw_item = log_raw.pivot_table(index='item_topic', columns='event', values='source_topic', aggfunc='count')\
                                                            .reset_index().sort_values(by='show', ascending=False)
log_raw_item

event          item_topic  click   show  view
11                  Наука   1583  19729  1257
13              Отношения   1432  18760  1276
6        Интересные факты   1492  17797  1435
12               Общество   1585  17392  1441
14               Подборки   1241  16002  1200
19                 Россия   1260  15401  1016
15        Полезные советы   1149  13776  1132
8                 История   1162  13652  1177
20                  Семья    850  10787   762
3                 Женщины    949  10108   884
18               Рассказы    592  10051   687
1                    Дети    824   9872   754
4                Здоровье    802   9396   629
10               Культура    717   9265   637
0                  Деньги    866   9189   609
9                 Красота    748   8718   709
22                 Туризм    684   8607   589
21               Скандалы    716   8296   614
24                   Юмор    749   8252   744
17            Путешествия    677   8245   742
7               Искусство    587   7605   666
16             Психология    701   6894   758
5            Знаменитости    531   6827   337
2      Женская психология    706   6754   575
23                    Шоу    632   6594   557

In [12]:
fig = go.Figure()
fig.add_trace(go.Bar(
    y=log_raw_item['click'],
    x=log_raw_item['item_topic'],
    name='Карточка отображена для пользователя',
    marker_color='#3c40c6'
))
fig.add_trace(go.Bar(
    y=log_raw_item['show'],
    x=log_raw_item['item_topic'],
    name='Пользователь кликнул на карточку',
    marker_color='#0fbcf9'
))

fig.add_trace(go.Bar(
    y=log_raw_item['view'],
    x=log_raw_item['item_topic'],
    name='Пользователь просмотрел статью карточки',
    marker_color='#ff5e57'
))

fig.update_layout(barmode='group', xaxis_tickangle=-45,
        title={
        'text': 'Способы взаимодействия пользователей с системой',
        'y':0.85,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig.show()

Сгруппируем способы взаимодействия пользователей с системой для определения % соотношения:

In [23]:
events_rating = log_raw.groupby('event').agg({'user_id': 'count'})

events_rating['events'] = (log_raw.groupby("event").agg({'user_id':"count"})
         .apply(lambda x: 100*x/x.sum())).round(2)

events_rating.columns = ['количество', '% от общего кол-ва']
events_rating.reset_index().sort_values('% от общего кол-ва', ascending=False)

event  количество  % от общего кол-ва
1   show      277969               86.22
0  click       23235                7.21
2   view       21187                6.57

Показатель кликабельности

In [21]:
CTR = 23235/277969 * 100
CTR

8.358845770571538

In [22]:
fig = go.Figure(data=go.Pie(labels=events_rating['количество'].value_counts()
                            .reset_index()['index'], values=log_raw['event'].value_counts()))
fig.update_layout(
    title={
        'text':'Соотношение способов взимодействия',
        'y':0.87,
        'x':0.5})
for trace in fig.data:
    trace['labels']=['click', 'show', 'view']
fig.show()

C:\Users\skd23\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\numpy\core\numeric.py:2446: FutureWarning:

elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison




[ссылка на Tableau Public](https://public.tableau.com/app/profile/nikita.semenov/viz/Yandex_Zendashboard_16417409556140/dashboard_zen?publish=yes/)

